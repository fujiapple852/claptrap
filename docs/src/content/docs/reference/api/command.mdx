---
title: "Command"
description: "The Command API"
---

<div class="api-table">
| API | Description |
| --- | ----------- |
| [`args`](args) | Adds multiple arguments to the list of valid possibilities. |
| [`groups`](groups) | Adds multiple `ArgGroups` to the Command at once. |
| [`subcommands`](subcommands) | Adds multiple subcommands to the list of valid possibilities. |
| [`no_binary_name`](no_binary_name) | Specifies that the parser should not assume the first argument passed is the binary name. |
| [`ignore_errors`](ignore_errors) | Try not to fail on parse errors, like missing option values. |
| [`args_override_self`](args_override_self) | Replace prior occurrences of arguments rather than error. |
| [`dont_delimit_trailing_values`](dont_delimit_trailing_values) | Disables the automatic delimiting of values after `--` or when `Arg::trailing_var_arg` was used. |
| [`color`](color) | Sets when to color output. |
| [`styles`](styles) | Sets the `Styles` for terminal output. |
| [`term_width`](term_width) | Sets the terminal width at which to wrap help messages. |
| [`max_term_width`](max_term_width) | Limit the line length for wrapping help when using the current terminal’s width. |
| [`disable_version_flag`](disable_version_flag) | Disables -V and --version flag. |
| [`propagate_version`](propagate_version) | Specifies to use the version of the current command for all subcommands. |
| [`next_line_help`](next_line_help) | Places the help string for all arguments and subcommands on the line after them. |
| [`disable_help_flag`](disable_help_flag) | Disables `-h` and `--help` flag. |
| [`disable_help_subcommand`](disable_help_subcommand) | Disables the help subcommand. |
| [`disable_colored_help`](disable_colored_help) | Disables colorized help messages. |
| [`help_expected`](help_expected) | Panic if help descriptions are omitted. |
| [`hide_possible_values`](hide_possible_values) | Tells clap not to print possible values when displaying help information. |
| [`infer_long_args`](infer_long_args) | Allow partial matches of long arguments or their aliases. |
| [`infer_subcommands`](infer_subcommands) | Allow partial matches of subcommand names and their aliases. |
| [`name`](name) | Sets the program’s name. |
| [`bin_name`](bin_name) | Overrides the runtime-determined name of the binary for help and error messages. |
| [`display_name`](display_name) | Overrides the runtime-determined display name of the program for help and error messages. |
| [`author`](author) | Sets the author(s) for the help message. |
| [`about`](about) | Sets the program’s description for the short help (`-h`). |
| [`long_about`](long_about) | Sets the program’s description for the long help (`--help`). |
| [`after_help`](after_help) | Free-form help text for after auto-generated short help (`-h`). |
| [`after_long_help`](after_long_help) | Free-form help text for after auto-generated long help (`--help`). |
| [`before_help`](before_help) | Free-form help text for before auto-generated short help (`-h`). |
| [`before_long_help`](before_long_help) | Free-form help text for before auto-generated long help (`--help`). |
| [`version`](version) | Sets the version for the short version (`-V`) and help messages. |
| [`long_version`](long_version) | Sets the version for the long version (`--version`) and help messages. |
| [`override_usage`](override_usage) | Overrides the clap generated usage string for help and error messages. |
| [`override_help`](override_help) | Overrides the clap generated help message (both `-h` and `--help`). |
| [`help_template`](help_template) | Sets the help template to be used, overriding the default format. |
| [`flatten_help`](flatten_help) | Flatten subcommand help into the current command’s help. |
| [`next_help_heading`](next_help_heading) | Set the default section heading for future args. |
| [`next_display_order`](next_display_order) | Change the starting value for assigning future display orders for args. |
| [`arg_required_else_help`](arg_required_else_help) | Exit gracefully if no arguments are present (e.g. `$ myprog`). |
| [`allow_missing_positional`](allow_missing_positional) | Allows one to implement two styles of CLIs where positionals can be used out of order. |
| [`short_flag`](short_flag) | Sets the short version of the subcommand flag without the preceding `-`. |
| [`long_flag`](long_flag) | Sets the long version of the subcommand flag without the preceding `--`. |
| [`alias`](alias) | Sets a hidden alias to this subcommand. |
| [`short_flag_alias`](short_flag_alias) | Add an alias, which functions as “hidden” short flag subcommand. |
| [`long_flag_alias`](long_flag_alias) | Add an alias, which functions as a “hidden” long flag subcommand. |
| [`aliases`](aliases) | Sets multiple hidden aliases to this subcommand. |
| [`short_flag_aliases`](short_flag_aliases) | Add aliases, which function as “hidden” short flag subcommands. |
| [`long_flag_aliases`](long_flag_aliases) | Add aliases, which function as “hidden” long flag subcommands. |
| [`visible_alias`](visible_alias) | Sets a visible alias to this subcommand. |
| [`visible_short_flag_alias`](visible_short_flag_alias) | Add an alias, which functions as “visible” short flag subcommand. |
| [`visible_long_flag_alias`](visible_long_flag_alias) | Add an alias, which functions as a “visible” long flag subcommand. |
| [`visible_aliases`](visible_aliases) | Sets multiple visible aliases to this subcommand. |
| [`visible_short_flag_aliases`](visible_short_flag_aliases) | Add aliases, which function as visible short flag subcommands. |
| [`visible_long_flag_aliases`](visible_long_flag_aliases) | Add aliases, which function as visible long flag subcommands. |
| [`display_order`](display_order) | Set the placement of this subcommand within the help. |
| [`hide`](hide) | Specifies that this subcommand should be hidden from help messages. |
| [`subcommand_required`](subcommand_required) | If no subcommand is present at runtime, error and exit gracefully. |
| [`allow_external_subcommands`](allow_external_subcommands) | Assume unexpected positional arguments are a subcommand. |
| [`external_subcommand_value_parser`](external_subcommand_value_parser) | Specifies how to parse external subcommand arguments. |
| [`args_conflicts_with_subcommands`](args_conflicts_with_subcommands) | Specifies that use of an argument prevents the use of subcommands. |
| [`subcommand_precedence_over_arg`](subcommand_precedence_over_arg) | Prevent subcommands from being consumed as an arguments value. |
| [`subcommand_negates_reqs`](subcommand_negates_reqs) | Allows subcommands to override all requirements of the parent command. |
| [`multicall`](multicall) | Multiple-personality program dispatched on the binary name (`argv[0]`). |
| [`subcommand_value_name`](subcommand_value_name) | Sets the value name used for subcommands when printing usage and help. |
| [`subcommand_help_heading`](subcommand_help_heading) | Sets the help heading used for subcommands when printing usage and help. |
</div>