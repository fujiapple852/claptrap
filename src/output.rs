use clap::builder::StyledStr;
use itertools::Itertools;

fn shell_quote(value: &str) -> String {
    format!("'{}'", value.replace('\'', "'\\''"))
}
use std::fmt::Display;

// The prefix for variables output by claptrap
const PREFIX: &str = "claptrap";

/// Represents the output of a claptrap command.
#[derive(Debug, Eq, PartialEq)]
pub enum Output {
    Cat(CatCmd),
    Variables(Vec<Var>),
}

impl Display for Output {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Cat(cmd) => write!(f, "{cmd}"),
            Self::Variables(vars) => write!(f, "{}", vars.iter().format("\n")),
        }
    }
}

/// Represents a variable output by claptrap.
#[derive(Debug, Eq, PartialEq)]
pub enum Var {
    Single(Vec<String>, String, String),
    Many(Vec<String>, String, Vec<String>),
}

impl Display for Var {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Single(prefix, name, value) => {
                let value = shell_quote(value);
                if prefix.is_empty() {
                    write!(f, "{PREFIX}_{name}={value}")
                } else {
                    write!(
                        f,
                        "{}_{}_{}={}",
                        PREFIX,
                        prefix.iter().format("_"),
                        name,
                        value
                    )
                }
            }
            Self::Many(prefix, name, values) => {
                let values = values.iter().map(|v| shell_quote(v)).join(" ");
                if prefix.is_empty() {
                    write!(f, "{PREFIX}_{name}=({values})")
                } else {
                    write!(
                        f,
                        "{}_{}_{}=({})",
                        PREFIX,
                        prefix.iter().format("_"),
                        name,
                        values
                    )
                }
            }
        }
    }
}

/// Exit code for the `CatCmd`.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ExitCode {
    Success = 0,
    Error = 1,
    Usage = 2,
    Panic = 3,
}

impl Display for ExitCode {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", *self as i32)
    }
}

/// A `cat` command invocation.
#[derive(Debug, Eq, PartialEq)]
pub struct CatCmd {
    pub data: StyledStr,
    pub exit_code: ExitCode,
}

impl CatCmd {
    #[must_use]
    pub fn new(cmd: StyledStr, exit_code: ExitCode) -> Self {
        Self {
            data: cmd,
            exit_code,
        }
    }
}

impl Display for CatCmd {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "command cat <<'EOF'\n{}EOF\nexit {}",
            self.data.ansi(),
            self.exit_code
        )
    }
}
