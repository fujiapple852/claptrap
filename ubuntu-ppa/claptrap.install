target/release/claptrap usr/bin
