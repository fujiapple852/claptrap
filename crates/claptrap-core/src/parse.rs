use crate::clap_ext::IsManyEx;
use crate::output::{CatCmd, ExitCode, Output, Var};
use crate::types::{Command, ValueParser};
use clap::parser::ValuesRef;
use clap::{ArgAction, ColorChoice};
use std::ffi::OsString;
use std::fmt::Display;

/// Parse the provided arguments and generate output.
///
/// This function does not perform any I/O operations.
#[must_use]
#[expect(clippy::needless_pass_by_value)]
pub fn parse<I, T>(cmd: Command, args: I) -> Output
where
    I: IntoIterator<Item = T>,
    T: Into<OsString> + Clone,
{
    let clap_cmd = clap::Command::from(cmd.clone());
    let color = clap_cmd.get_color();
    let help_color = if clap_cmd.is_disable_colored_help_set() {
        ColorChoice::Never
    } else {
        color
    };
    match clap_cmd.clone().try_get_matches_from(args) {
        Ok(matches) => Output::Variables(
            extract_subcommand_path(&matches)
                .into_iter()
                .chain(extract_matches(&cmd, &clap_cmd, &matches, &[]))
                .collect(),
        ),
        Err(err) => match err.kind() {
            clap::error::ErrorKind::DisplayHelp | clap::error::ErrorKind::DisplayVersion => {
                Output::Cat(CatCmd::new(err.render(), ExitCode::Success, help_color))
            }
            clap::error::ErrorKind::DisplayHelpOnMissingArgumentOrSubcommand => {
                Output::Cat(CatCmd::new(err.render(), ExitCode::Usage, help_color))
            }
            _ => Output::Cat(CatCmd::new(err.render(), ExitCode::Error, color)),
        },
    }
}

fn extract_subcommand_path(matches: &clap::ArgMatches) -> Option<Var> {
    let path = std::iter::successors(matches.subcommand(), |&(_, sub)| sub.subcommand())
        .map(|(name, _)| name.to_string())
        .collect::<Vec<_>>();
    (!path.is_empty()).then_some(Var::Subcommand(path))
}

fn extract_matches(
    cmd: &Command,
    clap_cmd: &clap::Command,
    matches: &clap::ArgMatches,
    prefix: &[String],
) -> Vec<Var> {
    let vars = matches.ids().filter_map(|id| {
        let arg = cmd.get_arguments().find(|a| a.get_id() == id.as_str())?;
        let clap_arg = clap_cmd.get_arguments().find(|a| a.get_id() == id)?;
        match clap_arg.get_action() {
            ArgAction::SetTrue | ArgAction::SetFalse => Some(Var::Single(
                prefix.to_vec(),
                id.to_string(),
                matches.get_flag(id.as_str()).to_string(),
            )),
            ArgAction::Count => Some(Var::Single(
                prefix.to_vec(),
                id.to_string(),
                matches.get_count(id.as_str()).to_string(),
            )),
            ArgAction::Append => arg
                .value_parser()
                .and_then(|v| get_many(v, matches, id.as_str()))
                .or_else(|| get_many_raw(matches, id.as_str()))
                .map(|vals| Var::Many(prefix.to_vec(), id.to_string(), vals)),
            ArgAction::Set => {
                if clap_arg.is_many() {
                    arg.value_parser()
                        .and_then(|v| get_many(v, matches, id.as_str()))
                        .or_else(|| get_many_raw(matches, id.as_str()))
                        .map(|vals| Var::Many(prefix.to_vec(), id.to_string(), vals))
                } else {
                    arg.value_parser()
                        .and_then(|v| get_one(v, matches, id.as_str()))
                        .or_else(|| get_one_raw(matches, id.as_str()))
                        .map(|val| Var::Single(prefix.to_vec(), id.to_string(), val))
                }
            }
            _ => None,
        }
    });

    let sub = matches.subcommand().into_iter().flat_map(|(name, sub_m)| {
        cmd.get_subcommands()
            .find(|sc| sc.get_name() == name)
            .into_iter()
            .filter_map(move |sub_cmd| {
                clap_cmd
                    .get_subcommands()
                    .find(|sc| sc.get_name() == name)
                    .map(|clap_sub| {
                        let mut sub_prefix = prefix.to_vec();
                        sub_prefix.push(name.to_string());
                        extract_matches(sub_cmd, clap_sub, sub_m, &sub_prefix)
                    })
            })
            .flatten()
    });

    vars.chain(sub).collect()
}

fn get_one(value_parser: &ValueParser, matches: &clap::ArgMatches, id: &str) -> Option<String> {
    match value_parser {
        ValueParser::Falsey | ValueParser::Boolish | ValueParser::Bool => {
            matches.get_one::<bool>(id).map(ToString::to_string)
        }
        ValueParser::String | ValueParser::PossibleValues(_) => {
            matches.get_one::<String>(id).map(ToString::to_string)
        }
        ValueParser::U8 => matches.get_one::<u8>(id).map(ToString::to_string),
        ValueParser::U16 => matches.get_one::<u16>(id).map(ToString::to_string),
        ValueParser::U32 => matches.get_one::<u32>(id).map(ToString::to_string),
        ValueParser::U64 => matches.get_one::<u64>(id).map(ToString::to_string),
        ValueParser::U128 => matches.get_one::<u128>(id).map(ToString::to_string),
        ValueParser::Usize => matches.get_one::<usize>(id).map(ToString::to_string),
        ValueParser::I8 => matches.get_one::<i8>(id).map(ToString::to_string),
        ValueParser::I16 => matches.get_one::<i16>(id).map(ToString::to_string),
        ValueParser::I32 => matches.get_one::<i32>(id).map(ToString::to_string),
        ValueParser::I64 => matches.get_one::<i64>(id).map(ToString::to_string),
        ValueParser::I128 => matches.get_one::<i128>(id).map(ToString::to_string),
        ValueParser::Isize => matches.get_one::<isize>(id).map(ToString::to_string),
        ValueParser::F32 => matches.get_one::<f32>(id).map(ToString::to_string),
        ValueParser::F64 => matches.get_one::<f64>(id).map(ToString::to_string),
    }
}

fn get_many(
    value_parser: &ValueParser,
    matches: &clap::ArgMatches,
    id: &str,
) -> Option<Vec<String>> {
    fn to_string<'a, T: Display + 'a>(values: ValuesRef<'a, T>) -> Vec<String> {
        values.map(ToString::to_string).collect()
    }
    match value_parser {
        ValueParser::Falsey | ValueParser::Boolish | ValueParser::Bool => {
            matches.get_many::<bool>(id).map(to_string)
        }
        ValueParser::String | ValueParser::PossibleValues(_) => {
            matches.get_many::<String>(id).map(to_string)
        }
        ValueParser::U8 => matches.get_many::<u8>(id).map(to_string),
        ValueParser::U16 => matches.get_many::<u16>(id).map(to_string),
        ValueParser::U32 => matches.get_many::<u32>(id).map(to_string),
        ValueParser::U64 => matches.get_many::<u64>(id).map(to_string),
        ValueParser::U128 => matches.get_many::<u128>(id).map(to_string),
        ValueParser::Usize => matches.get_many::<usize>(id).map(to_string),
        ValueParser::I8 => matches.get_many::<i8>(id).map(to_string),
        ValueParser::I16 => matches.get_many::<i16>(id).map(to_string),
        ValueParser::I32 => matches.get_many::<i32>(id).map(to_string),
        ValueParser::I64 => matches.get_many::<i64>(id).map(to_string),
        ValueParser::I128 => matches.get_many::<i128>(id).map(to_string),
        ValueParser::Isize => matches.get_many::<isize>(id).map(to_string),
        ValueParser::F32 => matches.get_many::<f32>(id).map(to_string),
        ValueParser::F64 => matches.get_many::<f64>(id).map(to_string),
    }
}

fn get_one_raw(matches: &clap::ArgMatches, id: &str) -> Option<String> {
    matches
        .get_raw(id)
        .and_then(|mut values| values.next())
        .map(|value| value.to_string_lossy().into_owned())
}

fn get_many_raw(matches: &clap::ArgMatches, id: &str) -> Option<Vec<String>> {
    matches
        .get_raw(id)
        .map(|values| values.map(|v| v.to_string_lossy().into_owned()).collect())
}
