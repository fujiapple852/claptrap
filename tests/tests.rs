use claptrap::command::Command;
use claptrap::parse;
use std::ffi::OsString;

macro_rules! case {
    ($path:expr, $name:expr) => {{
        (
            $name,
            include_str!(concat!("resources/", $path, "/", $name, ".toml")),
            include_str!(concat!("resources/", $path, "/", $name, ".args")),
        )
    }};
}

fn run(spec: &str, args: &str) -> claptrap::output::Output {
    let app: Command = toml::from_str(spec).unwrap();
    let args: Vec<OsString> = if args.trim().is_empty() {
        Vec::new()
    } else {
        args.split_whitespace().map(OsString::from).collect()
    };
    parse(app, args)
}

mod basic {
    use claptrap::output::{CatCmd, ExitCode};
    use test_case::test_case;

    #[test_case(case!("basic", "no_args"))]
    #[test_case(case!("basic", "short_help"))]
    #[test_case(case!("basic", "long_help"))]
    #[test_case(case!("basic", "short_version"))]
    #[test_case(case!("basic", "variables"))]
    #[test_case(case!("basic", "unexpected_arg"))]
    fn test_values((name, spec, args): (&str, &str, &str)) {
        insta::assert_snapshot!(name, super::run(spec, args));
    }

    #[test]
    fn cat_cmd_handles_eof_in_message() {
        let styled =
            clap::builder::StyledStr::from("this contains EOF in the text\nEOF\nand more\n");
        let cmd = CatCmd::new(styled, ExitCode::Error);
        insta::assert_snapshot!(format!("{}", cmd));
    }
}

mod command {
    use test_case::test_case;

    #[test_case(case!("command", "no_args"))]
    #[test_case(case!("command", "arg"))]
    #[test_case(case!("command", "arg_required_else_help"))]
    #[test_case(case!("command", "group"))]
    #[test_case(case!("command", "groups"))]
    #[test_case(case!("command", "subcommand"))]
    #[test_case(case!("command", "subcommand_many"))]
    #[test_case(case!("command", "ignore_error"))]
    #[test_case(case!("command", "args_override_self"))]
    #[test_case(case!("command", "dont_delimit_trailing_values"))]
    #[test_case(case!("command", "color"))]
    #[test_case(case!("command", "styles"))]
    #[test_case(case!("command", "term_width"))]
    #[test_case(case!("command", "max_term_width"))]
    #[test_case(case!("command", "disable_version_flag-1"))]
    #[test_case(case!("command", "disable_version_flag-2"))]
    #[test_case(case!("command", "propagate_version"))]
    #[test_case(case!("command", "next_line_help"))]
    #[test_case(case!("command", "disable_help_flag-1"))]
    #[test_case(case!("command", "disable_help_flag-2"))]
    #[test_case(case!("command", "disable_help_subcommand"))]
    #[test_case(case!("command", "disable_colored_help"))]
    #[test_case(case!("command", "help_expected"))]
    #[test_case(case!("command", "hide_possible_values"))]
    #[test_case(case!("command", "infer_long_args"))]
    #[test_case(case!("command", "infer_long_args_ambiguous"))]
    #[test_case(case!("command", "infer_subcommands"))]
    #[test_case(case!("command", "bin_name"))]
    #[test_case(case!("command", "display_name"))]
    #[test_case(case!("command", "author"))]
    #[test_case(case!("command", "about"))]
    #[test_case(case!("command", "after_help"))]
    #[test_case(case!("command", "before_help"))]
    #[test_case(case!("command", "before_long_help"))]
    #[test_case(case!("command", "version"))]
    #[test_case(case!("command", "long_version"))]
    #[test_case(case!("command", "override_usage"))]
    #[test_case(case!("command", "override_help"))]
    #[test_case(case!("command", "template"))]
    #[test_case(case!("command", "flatten_help"))]
    #[test_case(case!("command", "next_help_heading"))]
    #[test_case(case!("command", "next_display_order"))]
    #[test_case(case!("command", "allow_missing_positional"))]
    #[test_case(case!("command", "allow_missing_positional_with_default"))]
    #[test_case(case!("command", "allow_missing_positional_style_2-1"))]
    #[test_case(case!("command", "allow_missing_positional_style_2-2"))]
    #[test_case(case!("command", "short_flag"))]
    #[test_case(case!("command", "long_flag"))]
    #[test_case(case!("command", "alias"))]
    #[test_case(case!("command", "short_flag_alias"))]
    #[test_case(case!("command", "long_flag_alias"))]
    #[test_case(case!("command", "aliases"))]
    #[test_case(case!("command", "short_flag_aliases"))]
    #[test_case(case!("command", "long_flag_aliases"))]
    #[test_case(case!("command", "visible_alias"))]
    #[test_case(case!("command", "visible_short_flag_alias"))]
    #[test_case(case!("command", "visible_long_flag_alias"))]
    #[test_case(case!("command", "visible_aliases"))]
    #[test_case(case!("command", "visible_short_flag_aliases"))]
    #[test_case(case!("command", "visible_long_flag_aliases"))]
    #[test_case(case!("command", "display_order"))]
    #[test_case(case!("command", "hide"))]
    #[test_case(case!("command", "subcommand_required"))]
    #[test_case(case!("command", "allow_external_subcommands"))]
    #[test_case(case!("command", "external_subcommand_value_parser"))]
    #[test_case(case!("command", "args_conflicts_with_subcommands"))]
    #[test_case(case!("command", "subcommand_precedence_over_arg"))]
    #[test_case(case!("command", "subcommand_negates_reqs"))]
    #[test_case(case!("command", "subcommand_value_name"))]
    #[test_case(case!("command", "subcommand_help_heading"))]
    fn test_command((name, spec, args): (&str, &str, &str)) {
        insta::assert_snapshot!(name, super::run(spec, args));
    }

    #[test]
    #[ignore = "test is sensitive to terminal width"]
    fn long_about() {
        let output = super::run(
            include_str!("resources/command/long_about.toml"),
            include_str!("resources/command/long_about.args"),
        );
        insta::assert_snapshot!("long_about", output);
    }

    #[test]
    #[ignore = "test is sensitive to terminal width"]
    fn after_long_help() {
        let output = super::run(
            include_str!("resources/command/after_long_help.toml"),
            include_str!("resources/command/after_long_help.args"),
        );
        insta::assert_snapshot!("after_long_help", output);
    }
}

mod arg {
    use test_case::test_case;

    #[test_case(case!("arg", "action_append"))]
    #[test_case(case!("arg", "action_count"))]
    #[test_case(case!("arg", "action_help"))]
    #[test_case(case!("arg", "action_help_long"))]
    #[test_case(case!("arg", "action_help_short"))]
    #[test_case(case!("arg", "action_set"))]
    #[test_case(case!("arg", "action_set_false"))]
    #[test_case(case!("arg", "action_set_true"))]
    #[test_case(case!("arg", "action_version"))]
    #[test_case(case!("arg", "alias"))]
    #[test_case(case!("arg", "aliases"))]
    #[test_case(case!("arg", "allow_hyphen_values"))]
    #[test_case(case!("arg", "allow_negative_numbers"))]
    #[test_case(case!("arg", "conflicts_with"))]
    #[test_case(case!("arg", "conflicts_with_all"))]
    #[test_case(case!("arg", "default_missing_value-1"))]
    #[test_case(case!("arg", "default_missing_value-2"))]
    #[test_case(case!("arg", "default_missing_value-3"))]
    #[test_case(case!("arg", "default_missing_value_bool_literal-1"))]
    #[test_case(case!("arg", "default_missing_value_bool_literal-2"))]
    #[test_case(case!("arg", "default_missing_value_bool_literal-3"))]
    #[test_case(case!("arg", "default_value-1"))]
    #[test_case(case!("arg", "default_value-2"))]
    #[test_case(case!("arg", "default_value_if_equals-1"))]
    #[test_case(case!("arg", "default_value_if_equals-2"))]
    #[test_case(case!("arg", "default_value_if_not_present"))]
    #[test_case(case!("arg", "default_value_if_present"))]
    #[test_case(case!("arg", "default_value_if_unset"))]
    #[test_case(case!("arg", "default_value_ifs-1"))]
    #[test_case(case!("arg", "default_value_ifs-2"))]
    #[test_case(case!("arg", "default_value_ifs_order"))]
    #[test_case(case!("arg", "default_values-1"))]
    #[test_case(case!("arg", "default_values-2"))]
    #[test_case(case!("arg", "display_order"))]
    #[test_case(case!("arg", "exclusive"))]
    #[test_case(case!("arg", "group"))]
    #[test_case(case!("arg", "groups"))]
    #[test_case(case!("arg", "help"))]
    #[test_case(case!("arg", "help_heading"))]
    #[test_case(case!("arg", "hide"))]
    #[test_case(case!("arg", "hide_default_value"))]
    #[test_case(case!("arg", "hide_env"))]
    #[test_case(case!("arg", "hide_env_values"))]
    #[test_case(case!("arg", "hide_long_help-1"))]
    #[test_case(case!("arg", "hide_long_help-2"))]
    #[test_case(case!("arg", "hide_possible_values"))]
    #[test_case(case!("arg", "hide_short_help-1"))]
    #[test_case(case!("arg", "hide_short_help-2"))]
    #[test_case(case!("arg", "ignore_case"))]
    #[test_case(case!("arg", "ignore_case_multi"))]
    #[test_case(case!("arg", "index"))]
    #[test_case(case!("arg", "it_fails_when_action_set_false_twice"))]
    #[test_case(case!("arg", "it_fails_when_action_set_true_twice"))]
    #[test_case(case!("arg", "it_fails_when_action_set_twice"))]
    #[test_case(case!("arg", "it_fails_when_not_allow_hyphen_values"))]
    #[test_case(case!("arg", "last-1"))]
    #[test_case(case!("arg", "last-2"))]
    #[test_case(case!("arg", "long"))]
    #[test_case(case!("arg", "next_line_help"))]
    #[test_case(case!("arg", "num_args"))]
    #[test_case(case!("arg", "num_args_default_missing_value-1"))]
    #[test_case(case!("arg", "num_args_default_missing_value-2"))]
    #[test_case(case!("arg", "num_args_default_missing_value-3"))]
    #[test_case(case!("arg", "num_args_multi_positional-1"))]
    #[test_case(case!("arg", "num_args_multi_positional-2"))]
    #[test_case(case!("arg", "num_args_multi_positional_solution"))]
    #[test_case(case!("arg", "num_args_tuples-1"))]
    #[test_case(case!("arg", "num_args_tuples-2"))]
    #[test_case(case!("arg", "overrides_with"))]
    #[test_case(case!("arg", "overrides_with_all"))]
    #[test_case(case!("arg", "require_equals-1"))]
    #[test_case(case!("arg", "require_equals-2"))]
    #[test_case(case!("arg", "required-1"))]
    #[test_case(case!("arg", "required-2"))]
    #[test_case(case!("arg", "required_if_eq-1"))]
    #[test_case(case!("arg", "required_if_eq-2"))]
    #[test_case(case!("arg", "required_if_eq_all-1"))]
    #[test_case(case!("arg", "required_if_eq_all-2"))]
    #[test_case(case!("arg", "required_if_eq_any-1"))]
    #[test_case(case!("arg", "required_if_eq_any-2"))]
    #[test_case(case!("arg", "required_if_eq_case_insensitive"))]
    #[test_case(case!("arg", "required_if_eq_case_sensitive"))]
    #[test_case(case!("arg", "required_unless_present-1"))]
    #[test_case(case!("arg", "required_unless_present-2"))]
    #[test_case(case!("arg", "required_unless_present_all-1"))]
    #[test_case(case!("arg", "required_unless_present_all-2"))]
    #[test_case(case!("arg", "required_unless_present_any-1"))]
    #[test_case(case!("arg", "required_unless_present_any-2"))]
    #[test_case(case!("arg", "requires-1"))]
    #[test_case(case!("arg", "requires-2"))]
    #[test_case(case!("arg", "requires_if-1"))]
    #[test_case(case!("arg", "requires_if-2"))]
    #[test_case(case!("arg", "requires_ifs-1"))]
    #[test_case(case!("arg", "requires_ifs-2"))]
    #[test_case(case!("arg", "requires_ifs_present-1"))]
    #[test_case(case!("arg", "requires_ifs_present-2"))]
    #[test_case(case!("arg", "short"))]
    #[test_case(case!("arg", "short_alias"))]
    #[test_case(case!("arg", "short_aliases"))]
    #[test_case(case!("arg", "short_help"))]
    #[test_case(case!("arg", "trailing_var_arg"))]
    #[test_case(case!("arg", "value_delimiter"))]
    #[test_case(case!("arg", "value_name"))]
    #[test_case(case!("arg", "value_names"))]
    #[test_case(case!("arg", "value_parser"))]
    #[test_case(case!("arg", "value_terminator"))]
    #[test_case(case!("arg", "visible_alias"))]
    #[test_case(case!("arg", "visible_aliases"))]
    #[test_case(case!("arg", "visible_short_alias"))]
    #[test_case(case!("arg", "visible_short_aliases"))]
    fn test_arg((name, spec, args): (&str, &str, &str)) {
        insta::assert_snapshot!(name, super::run(spec, args));
    }

    #[test_case(case!("arg", "env"))]
    #[test_case(case!("arg", "env_default_value"))]
    #[test_case(case!("arg", "env_option"))]
    fn test_arg_env((name, spec, args): (&str, &str, &str)) {
        unsafe {
            std::env::set_var("MY_FLAG", "env");
        }
        insta::assert_snapshot!(name, super::run(spec, args));
    }

    #[test_case(case!("arg", "env_falsey"))]
    fn test_arg_env_falsey((name, spec, args): (&str, &str, &str)) {
        unsafe {
            std::env::set_var("TRUE_FLAG", "true");
            std::env::set_var("FALSE_FLAG", "0");
        }
        insta::assert_snapshot!(name, super::run(spec, args));
    }

    #[test_case(case!("arg", "env_multi"))]
    fn test_arg_env_multi((name, spec, args): (&str, &str, &str)) {
        unsafe {
            std::env::set_var("MY_FLAG_MULTI", "env1,env2");
        }
        insta::assert_snapshot!(name, super::run(spec, args));
    }

    #[test]
    #[ignore = "test is sensitive to terminal width"]
    fn long_help() {
        let output = super::run(
            include_str!("resources/arg/long_help.toml"),
            include_str!("resources/arg/long_help.args"),
        );
        insta::assert_snapshot!("long_help", output);
    }
}

mod arg_group {
    use test_case::test_case;

    #[test_case(case!("arg_group", "args"))]
    #[test_case(case!("arg_group", "multiple"))]
    #[test_case(case!("arg_group", "required"))]
    #[test_case(case!("arg_group", "requires"))]
    #[test_case(case!("arg_group", "requires_all"))]
    #[test_case(case!("arg_group", "conflicts_with"))]
    #[test_case(case!("arg_group", "conflicts_with_all"))]
    fn test_arg_group((name, spec, args): (&str, &str, &str)) {
        insta::assert_snapshot!(name, super::run(spec, args));
    }
}

mod values {
    use test_case::test_case;

    #[test_case(case!("values", "bool"))]
    #[test_case(case!("values", "boolish"))]
    #[test_case(case!("values", "falsey"))]
    #[test_case(case!("values", "string"))]
    #[test_case(case!("values", "u8"))]
    #[test_case(case!("values", "u16"))]
    #[test_case(case!("values", "u32"))]
    #[test_case(case!("values", "u64"))]
    #[test_case(case!("values", "u128"))]
    #[test_case(case!("values", "usize"))]
    #[test_case(case!("values", "i8"))]
    #[test_case(case!("values", "i16"))]
    #[test_case(case!("values", "i32"))]
    #[test_case(case!("values", "i64"))]
    #[test_case(case!("values", "i128"))]
    #[test_case(case!("values", "isize"))]
    #[test_case(case!("values", "f32"))]
    #[test_case(case!("values", "f64"))]
    #[test_case(case!("values", "possible_values"))]
    fn test_values((name, spec, args): (&str, &str, &str)) {
        insta::assert_snapshot!(name, super::run(spec, args));
    }
}

mod quoting {
    use test_case::test_case;

    #[test_case(case!("quoting", "single_value_with_quote"))]
    #[test_case(case!("quoting", "many_values_special_chars"))]
    fn test_quoting((name, spec, args): (&str, &str, &str)) {
        insta::assert_snapshot!(name, super::run(spec, args));
    }
}
