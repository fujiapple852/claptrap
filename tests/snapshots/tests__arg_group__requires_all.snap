---
source: tests/tests.rs
expression: "super::run(spec, args)"
---
command cat <<'EOF_bd3680b7'
[1m[31merror:[0m the following required arguments were not provided:
  [32m-v[0m

[1m[4mUsage:[0m [1mmyprog[0m [1m-d[0m [1m-v[0m [1m-c[0m

For more information, try '[1m--help[0m'.
EOF_bd3680b7
exit 1
